Globals: Repair dynamicTitle: false for user-defined tools

ia319 opened the pull request:
Closes #33281

<!-- If your PR is related to an issue, provide the number(s) above; if it resolves multiple issues, be sure to break them up (e.g. "closes #1000, closes #1001"). -->

<!--

Thank you for contributing to Storybook! Please submit all PRs to the `next` branch unless they are specific to the current release. Storybook maintainers cherry-pick bug and documentation fixes into the `main` branch as part of the release process, so you shouldn't need to worry about this. For additional guidance: https://storybook.js.org/docs/contribute

-->

## What I did

<!-- Briefly describe what your PR does -->
* Add `showSelectedLabel` prop to `Select` (default `true`).
* Only show the selected label when this prop is enabled.
* Set `showSelectedLabel={false}` in `ToolbarMenuSelect`.


## Checklist for Contributors

### Testing

<!-- Please check (put an "x" inside the "[ ]") the applicable items below to communicate how to test your changes -->

#### The changes in this PR are covered in the following automated tests:

- [ ] stories
- [ ] unit tests
- [ ] integration tests
- [ ] end-to-end tests

#### Manual testing

_This section is mandatory for all contributions. If you believe no manual test is necessary, please state so explicitly. Thanks!_

<!-- Please include the steps to test your changes here. For example:

1. Run a sandbox for template, e.g. `yarn task --task sandbox --start-from auto --template react-vite/default-ts`
2. Open Storybook in your browser
3. Access X story

-->

### Documentation

<!-- Please check (put an "x" inside the "[ ]") the applicable items below to indicate which documentation has been updated. -->

- [ ] Add or update documentation reflecting your changes
- [ ] If you are deprecating/removing a feature, make sure to update
      [MIGRATION.MD](https://github.com/storybookjs/storybook/blob/next/MIGRATION.md)

## Checklist for Maintainers

- [ ] When this PR is ready for testing, make sure to add `ci:normal`, `ci:merged` or `ci:daily` GH label to it to run a specific set of sandboxes. The particular set of sandboxes can be found in `code/lib/cli-storybook/src/sandbox-templates.ts`
- [ ] Make sure this PR contains **one** of the labels below:
   <details>
     <summary>Available labels</summary>

  - `bug`: Internal changes that fixes incorrect behavior.
  - `maintenance`: User-facing maintenance tasks.
  - `dependencies`: Upgrading (sometimes downgrading) dependencies.
  - `build`: Internal-facing build tooling & test updates. Will not show up in release changelog.
  - `cleanup`: Minor cleanup style change. Will not show up in release changelog.
  - `documentation`: Documentation **only** changes. Will not show up in release changelog.
  - `feature request`: Introducing a new feature.
  - `BREAKING CHANGE`: Changes that break compatibility in some way with current major version.
  - `other`: Changes that don't fit in the above categories.

   </details>

### 🦋 Canary release

<!-- CANARY_RELEASE_SECTION -->

This PR does not have a canary release associated. You can request a canary release of this pull request by mentioning the `@storybookjs/core` team here.

_core team members can create a canary release [here](https://github.com/storybookjs/storybook/actions/workflows/publish.yml) or locally with `gh workflow run --repo storybookjs/storybook publish.yml --field pr=<PR_NUMBER>`_

<!-- CANARY_RELEASE_SECTION -->

<!-- BENCHMARK_SECTION -->
<!-- BENCHMARK_SECTION -->


<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

* **New Features**
  * Added a configurable option to control whether selected option titles are displayed in select dropdowns. Default behavior is maintained for backward compatibility, automatically applying across single and multi-select variants.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

ia319 committed:
fix(core): add showSelectedLabel prop to Select to control selected label rendering

ia319 committed:
fix(core): disable Select selected-label override in ToolbarMenuSelect

ia319 committed:
fix lint error

ia319 committed:
refactor (ToolbarMenuSelect): remove manual title calculation, remove unused import, pass dynamicTitle via showSelectedLabel

ia319 committed:
Merge branch 'next' into bug/33281-dynamic-title-select

ia319 commented on code/core/src/toolbar/components/ToolbarMenuSelect.tsx:
Thanks for pointing that out — updated accordingly.

Sidnioulz commented on code/core/src/components/components/Select/Select.tsx:
We'll need new stories showing how labels behave on `!multiSelect` and on `multiSelect` Selects, when a selection is made, and the option is on or off (so, 4 stories with `defaultOptions` and a `play` function validating the HTML output).

Sidnioulz commented on code/core/src/toolbar/components/ToolbarMenuSelect.tsx:
Could you please make `dynamicTitle` `true` by default here?

Sidnioulz commented on code/core/src/toolbar/components/ToolbarMenuSelect.tsx:
```diff
@@ -40,16 +40,12 @@ export const ToolbarMenuSelect: FC<ToolbarMenuSelectProps> = withKeyboardCycle(
     const isOverridden = id in storyGlobals;
     let icon = _icon;
-    let title = _title;
+    const title = _title;
```
You no longer need to rename the `title` prop to `_title` and then create a new `title`. Could you please tidy that up?

Sidnioulz commented on code/core/src/toolbar/components/ToolbarMenuSelect.tsx:
Bunch more things to tidy up:

* [ ] In [docs/_snippets/storybook-preview-configure-globaltypes.md](https://github.com/storybookjs/storybook/blob/next/docs/_snippets/storybook-preview-configure-globaltypes.md), could you please replace the 4 occurences of `// Change title based on selected value` to `// Change title based on selected value (recommended for consistency with the Storybook UI)`
* [ ] Remove `getSelectedTitle`'s source code as it is no longer needed

Sidnioulz reviewed (changes requested):
Thanks for tackling this @ia319! This is a bit of a sensitive one as we wanted to enforce dynamic titles on all Selects (and yes, I did forget to handle that old option in ToolbarMenuSelect 😬).

ia319 committed:
refactor: rename legacy Select prop and clean up ToolbarMenuSelect title handling

- renamed `showSelectedLabel` to `showSelectedOptionTitle` and marked as legacy
- removed unused `_title` alias and redundant title reassignment
- enabled `dynamicTitle` by default

ia319 committed:
chore: update globaltypes snippet and remove unused getSelectedTitle

ia319 committed:
test(core): add four stories validating Select label behavior

added new stories demonstrating how labels render when:
- single-select: showSelectedOptionTitle = true / false
- multi-select: showSelectedOptionTitle = true / false

ia319 committed:
Merge branch 'bug/33281-dynamic-title-select' of https://github.com/ia319/storybook into bug/33281-dynamic-title-select

ia319 committed:
Merge branch 'next' into bug/33281-dynamic-title-select

Sidnioulz reviewed (commented):
Kudos @ia319, there's just one tiny thing to do with Select multiSelect and we can merge this. Thanks for your contribution ❤️

Sidnioulz committed:
Merge branch 'next' into bug/33281-dynamic-title-select

Sidnioulz reviewed (approved):
Thanks @ia319!

valentinpalkovic commented:
Hi @ia319 

Thank you for putting together this PR!

This PR has some merge conflicts. Would you be able to take a look?

ia319 committed:
Merge branch 'next' into bug/33281-dynamic-title-select

ia319 committed:
Merge branch 'next' into bug/33281-dynamic-title-select
